### PR TITLE
Add graceful shutdown to pgbouncer via entrypoint.sh

### PR DESCRIPTION
**Pull request by andriisoldatenko**
## Details

This PR catch SIGNTERM (15) and sends SIGNINT (2) to pgbouncer process.
It prevents to drop connections and immediate shutdown.

## Related Issues

- https://github.com/astronomer/airflow-operator/issues/1298
- https://github.com/astronomer/issues/issues/5358

## Checklist

- [x] version.txt was updated

Practically:
```
docker run p
Waiting for it to complete
+ trap cleanup TERM
+ main_proc=7
+ echo 'Waiting for it to complete'
+ wait
+ exec pgbouncer /etc/pgbouncer/pgbouncer.ini
2023-01-19 14:26:52.968 UTC [7] LOG kernel file descriptor limit: 1048576 (hard: 1048576); max_client_conn: 100, max expected fd use: 112
2023-01-19 14:26:52.969 UTC [7] LOG listening on 0.0.0.0:6432
2023-01-19 14:26:52.969 UTC [7] LOG listening on unix:/tmp/.s.PGSQL.6432
2023-01-19 14:26:52.969 UTC [7] LOG process up: PgBouncer 1.17.0, libevent 2.1.12-stable (epoll), adns: c-ares 1.18.1, tls: OpenSSL 1.1.1s  1 Nov 2022
```
and then run

```
docker kill --signal 15 62370a630b6f
```
you will get:
```
2023-01-19 14:26:52.968 UTC [7] LOG kernel file descriptor limit: 1048576 (hard: 1048576); max_client_conn: 100, max expected fd use: 112
2023-01-19 14:26:52.969 UTC [7] LOG listening on 0.0.0.0:6432
2023-01-19 14:26:52.969 UTC [7] LOG listening on unix:/tmp/.s.PGSQL.6432
2023-01-19 14:26:52.969 UTC [7] LOG process up: PgBouncer 1.17.0, libevent 2.1.12-stable (epoll), adns: c-ares 1.18.1, tls: OpenSSL 1.1.1s  1 Nov 2022
running cleanup
+ cleanup
+ echo running cleanup
+ kill -INT 7
+ wait 7
2023-01-19 14:27:32.598 UTC [7] LOG got SIGINT, shutting down
2023-01-19 14:27:32.614 UTC [7] LOG server connections dropped, exiting
exited with 0
+ echo exited with 0
```

### CODE REVIEW

File: pgbouncer/docker-entrypoint.sh [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+  kill -INT "$main_proc"
+  # The signal interrupted our `wait`, so lets do it again
+  wait "$main_proc"
+  echo exited with "$?"
+}
+# SIGTERM - Immediate shutdown. Same as issuing SHUTDOWN on the console.
+# but we would like to catch such signal and replace with SIGINT (Safe shutdown.)
+trap "cleanup" TERM
+
+exec "$@" &
```
Comment (ashb) [inline pgbouncer/docker-entrypoint.sh:15]: Does backgrounding exec actually work? I'd expect this to replace the bash process still.

```suggestion
# Run the real command in background so we can still respond to signals.
"$@" &
```

Comment (andriisoldatenko) [reply pgbouncer/docker-entrypoint.sh]: yes, check my example

Comment (andriisoldatenko) [reply pgbouncer/docker-entrypoint.sh]:
```
docker run p
Waiting for it to complete
+ trap cleanup TERM
+ main_proc=7
+ echo 'Waiting for it to complete'
+ wait
+ exec pgbouncer /etc/pgbouncer/pgbouncer.ini
2023-01-19 14:26:52.968 UTC [7] LOG kernel file descriptor limit: 1048576 (hard: 1048576); max_client_conn: 100, max expected fd use: 112
2023-01-19 14:26:52.969 UTC [7] LOG listening on 0.0.0.0:6432
2023-01-19 14:26:52.969 UTC [7] LOG listening on unix:/tmp/.s.PGSQL.6432
2023-01-19 14:26:52.969 UTC [7] LOG process up: PgBouncer 1.17.0, libevent 2.1.12-stable (epoll), adns: c-ares 1.18.1, tls: OpenSSL 1.1.1s  1 Nov 2022
running cleanup
+ cleanup
+ echo running cleanup
+ kill -INT 7
+ wait 7
2023-01-19 14:27:32.598 UTC [7] LOG got SIGINT, shutting down
2023-01-19 14:27:32.614 UTC [7] LOG server connections dropped, exiting
exited with 0
+ echo exited with 0
```

Comment (andriisoldatenko) [reply pgbouncer/docker-entrypoint.sh]: @ashb FYI: it works with `exec` and without.

**Review by ashb — APPROVED** [review]

File: pgbouncer/docker-entrypoint.sh [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  echo exited with "$?"
+}
+# SIGTERM - Immediate shutdown for pgbouncer process,
+# but we would like to catch such signal and replace with SIGINT (Safe shutdown.)
+trap "cleanup" TERM
+
+# Run the real command in background so we can still respond to signals.
+"$@" &
+main_proc=$!
+echo "Waiting for it to complete"
```
Comment (danielhoherd) [inline pgbouncer/docker-entrypoint.sh:18]:
```suggestion
echo "Watching pid $! $*"
```

File: pgbouncer/docker-entrypoint.sh [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+cleanup() {
+  echo running cleanup
+  kill -INT "$main_proc"
+  # The signal interrupted our `wait`, so lets do it again
+  wait "$main_proc"
+  echo exited with "$?"
+}
+# SIGTERM - Immediate shutdown for pgbouncer process,
+# but we would like to catch such signal and replace with SIGINT (Safe shutdown.)
+trap "cleanup" TERM
```
Comment (danielhoherd) [inline pgbouncer/docker-entrypoint.sh:13]:
```suggestion
trap "cleanup" TERM INT
```

**Comment by danielhoherd** [comment]
Are we OK with this entrypoint only catching the TERM signal and handling it? Do we want to pass any other signals through? I left a suggestion to handle INT too.

File: pgbouncer/docker-entrypoint.sh [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+cleanup() {
+  echo running cleanup
+  kill -INT "$main_proc"
+  # The signal interrupted our `wait`, so lets do it again
+  wait "$main_proc"
+  echo exited with "$?"
+}
+# SIGTERM - Immediate shutdown for pgbouncer process,
+# but we would like to catch such signal and replace with SIGINT (Safe shutdown.)
```
Comment (danielhoherd) [inline pgbouncer/docker-entrypoint.sh:12]: This comment should be moved to the top of the file.
```suggestion
# SIGTERM - Immediate shutdown for pgbouncer process,
# but we would like to catch such signal and replace with SIGINT (Safe shutdown.)
```

File: pgbouncer/docker-entrypoint.sh [hunk 5]
```diff
@@ -0,0 +1,19 @@
+#!/bin/sh
+set -ex
```
Comment (danielhoherd) [inline pgbouncer/docker-entrypoint.sh:2]:
```suggestion
#!/bin/sh
# Catch SIGTERM (immediate shutdown) and replace it with INT (safe shutdown)
# so pgbouncer can signal postgres clients that it is shutting down.
set -ex
```

**Comment by andriisoldatenko** [comment]
> Are we OK with this entrypoint only catching the TERM signal and handling it? Do we want to pass any other signals through? I left a suggestion to handle INT too.

thanks @danielhoherd fixed!

File: pgbouncer/docker-entrypoint.sh [hunk 6]
```diff
@@ -0,0 +1,22 @@
+#!/bin/sh
+# Catch SIGTERM (immediate shutdown) and replace it with INT (safe shutdown)
+# so pgbouncer can signal postgres clients that it is shutting down.
+#
+# SIGTERM - Immediate shutdown for pgbouncer process,
+# but we would like to catch such signal and replace with SIGINT (Safe shutdown.)
```
Comment (danielhoherd) [inline pgbouncer/docker-entrypoint.sh:6]: We can remove this if we're OK with the above comment.

**Review by danielhoherd — APPROVED** [review]
LGTM! Passing just the few signals we chose seems fine to me. I did some testing and passing them all can lead to a variety of other failure scenarios.

**Pull request merged**